File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.go
func findFirstOccurence(nums []int, target int, n int) int{
    
    first_occurence := -1
    low, high := 0, n-1
    
    for low <= high {
        
        mid := (low + high) >> 1
        
        if nums[mid] == target {
            first_occurence = mid
            high = mid - 1
        } else if nums[mid] > target {
            high = mid - 1
        } else {
            low = mid + 1
        }
    }
    
    return first_occurence
}

func findLastOccurence(nums []int, target int, n int) int{
    
    last_occurence := -1
    low, high := 0, n-1
    
    for low <= high {
        
        mid := (low + high) >> 1
        
        if nums[mid] == target {
            last_occurence = mid
            low = mid + 1
        } else if nums[mid] > target {
            high = mid - 1
        } else {
            low = mid + 1
        }
    }
    
    return last_occurence
}

func searchRange(nums []int, target int) []int {
    
    n := len(nums)
    
    firstOccurence := findFirstOccurence(nums, target, n)
    if firstOccurence == -1 {
        return []int {-1, -1}
    }
    
    lastOccurence := findLastOccurence(nums, target, n)
    
    return []int{firstOccurence, lastOccurence}
}

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
public:
    int lowerBound(vector<int>& nums, int target, int n){
        
        int lower_bound = n;
        int low = 0;
        int high = n-1;
        
        while(low <= high){
            
            int mid = (low + high) >> 1;
            
            if(nums[mid] >= target){
                lower_bound=mid;
                high = mid-1;
            }
            else{
                low = mid + 1;
            }
        }
        
        return lower_bound;
    }
    
    int upperBound(vector<int>& nums, int target, int n){
        
        int upper_bound = n;
        int low = 0;
        int high = n-1;
        
        while(low <= high){
            
            int mid = (low + high) >> 1;
            
            if(nums[mid] > target){
                upper_bound = mid;
                high = mid - 1;
            }
            else{
                low = mid + 1;
            }
        }
        
        return upper_bound;
    }
    
    vector<int> searchRange(vector<int>& nums, int target) {
       
        int n = nums.size();
        
        int lb = lowerBound(nums, target, n);
        int ub = upperBound(nums, target, n);
        
        if(lb == n || nums[lb] != target)
            return {-1, -1};
        
        return {lb, ub-1};
    }
};